### ISSUE TITLE
ElementHelper 2.0

### ISSUE THREAD

**Issue by roryg** [roryg/ElementHelper#31]
ElementHelper 2 is now up on the 2.0 branch, it's not quite finished yet but you can download a transport package for it [here](https://db.tt/w4k1jdYL) (tested and working on Revo 2.3.1). The code has been completely reworked/refactored and I think is much easier to work with now.

One of the biggest differences feature wise is that elements will now be synced both ways, so if you create an element in the manager a static file will be created for the element and vice versa. (This doesn't work for template variable elements to the template_variables.json file yet.)

There are a number of things still left to do:
1. Template variables need to be saved/removed into/from the template_variables.json file when created/deleted in the manager.
2. Add ability to create MIGx template variables
3. Create tests!
4. Decide if a setting for tv template access is needed.
5. Decide if a setting for auto removing elements is needed.
6. Since elements are synced both ways now there needs to be a way to prevent certain elements from being synced (e.g. a snippet or plugin that's installed with an extra).
7. And probably a bunch of other things I'm forgetting! Feel free to discuss ideas, thoughts etc for 2.0 here.

**Comment by kevinhamil** [comment]
@roryg I've just recently switched to a MODx setup and came across EH when searching best workflows... seeing the last discussion was 7 months ago, I was curious to find out if you've had any additional time to further develop this? I have the current version running and I love it, but of course, I am looking forward to a production release. Thanks.

**Comment by roryg** [comment]
@kevinhamil Yikes time flies! Apologies to all for neglecting this so badly, I'm definitely not abandoning it. I have projects coming up that I'll *need* ElementHelper for so I will be working on it soon.

**Comment by roryg** [comment]
Minor update available here - https://db.tt/SBKhONkV

Template Variables will now sync both ways including MIGX tvs. Unfortunately I haven't come up with a good way to sync template variable changes that have been made in the manager yet as Modx doesn't store a timestamp for them, but new TVs made in the manager will be synced to the TV json file.

**Comment by kevinhamil** [comment]
@roryg awesome, great news! Thanks.

**Comment by kevinhamil** [comment]
I don't know if this is the place to post a question like this, but I'm not sure where else it would be relevant to ask. In the template_variable.json, how would I define a specific Media Source? I can't seem to find how this would be included, or if it can be.

Something like...
`"media_source": ["myTVMediaSource"]`

If there's a more appropriate place to ask questions like this let me know. Thanks.

**Comment by roryg** [comment]
@kevinhamil There's no way to do that yet but it's definitely a feature I want to include.

**Comment by roryg** [comment]
Updated package available here - https://www.dropbox.com/s/8nn1s6k162ynov6/elementhelper-2.0.0-alpha-2.transport.zip?dl=0

I've added a setting for controlling template variable access and fixed the installation errors people were having. I've also added a category whitelist setting which allows you to whitelist categories for ElementHelper to work with. This should help prevent tracking elements that have been created by other Modx extras.

**Comment by zurie** [comment]
so i had ElementHelper 1.3 running o MODX 2.3.5 (not working). installed the latest alpha and now it unable to even access the login page.  it definetally works, because its generating TONS of stuff (every chunk i've ever made ever) but somethingis causing the site to error 500.  Any suggestions? (this is all localhost) maybe I can go in with phpadmin and disable it or something to fix it?

**Comment by DESIGNfromWITHIN** [comment]
Any update on this? Can it be made available from within MODX?

**Comment by roryg** [comment]
It's almost ready to be available within Modx but there was an issue reported here - https://github.com/roryg/ElementHelper/issues/38 - that I'll need to work out first. Hopefully before the year ends!

**Comment by jchirschy** [comment]
Hello @roryg,

I definitely second th request from @kevinhamil.
I have a huge using of media source for TV. It's a convenient way to guide the user through the tree of files. Would that feature be available with this new release ?

**Comment by roryg** [comment]
@jchirschy Yes, it's actually already available in the latest alpha version which you can find here - https://www.dropbox.com/s/5x627zz1cnq1rxw/elementhelper-2.0.0-alpha-3.transport.zip?dl=0 - it probably needs further testing but should hopefully work :)

Using it you add a media source to your TV json with the `media_source` property e.g. `"media_source": "Uploads"`.

**Comment by Izacktly** [comment]
Hello @roryg I must say this is a great plugin.  I installed from the test build posted above for Modx 2.4.2  However, the error log reports (ERROR @ /manager/index.php) [OnManagerPageInit].  and right after this the log prints the entire code from the articles plugin.  If I remove the articles extra, it then starts reporting on the Ace plugin and so on.  Anyone experiencing this?

**Comment by roryg** [comment]
Hi @Izacktly try adding the plugins causing the issues to the Element Blacklist setting, that should prevent ElementHelper from trying to process them. You might need to uninstall and reinstall the plugins after you've done that.

**Comment by Izacktly** [comment]
Understood.  Thank you very much for the quick response!
